Add support for gnomAD GRCh38 annotations

themisteryang opened the pull request:
- This PR adds support for gnomAD GRCh38 annotations in addition to the current support for gnomAD GRCh37 annotations.
- Going forward, all gnomAD annotations will be organized into separate collections in MongoDB, more specifically:
    - `GRCh37ExomeAnnotations`: for GRCh37 exome annotations;
    - `GRCh37GenomeAnnotations`: for GRCh37 genome annotations; and
    - `GRCh38GenomeAnnotations_chr<chr>`: a separate collection for each of the GRCh38 genome annotation chromosomes.
 - At this time, all gnomAD annotations has been imported into the production MongoDB instance and indexed for use on staging and production.
 - The legacy `annotations` collection will be retained until the changes from this PR have also been merged into `production`, after which it can be dropped.

Closes #269

themisteryang committed:
Replace the "gnomad_exome_AF" column with the "gnomAD_AF" column

- Add query for gnomAD GRCh37 genome annotations
- Calculate the gnomAD allele frequency as the highest value between the gnomAD exome allele frequency and the gnomAD genome allele frequency
- Rename the "gnomad_exome_AF" column to "gnomAD_AF"
- Update the header tooltip text for the "gnomAD_AF" column to specify how the gnomAD allele frequency is calculated

themisteryang committed:
Update annotateGnomad test

themisteryang committed:
Merge branch 'develop' into kevin/gnomAD_AF

themisteryang committed:
Run yarn format

themisteryang committed:
Fix ESLint error

themisteryang committed:
Add support for gnomAD GRCh38 annotations

- Account for the splitting of the annotations collection into separate collections (one for GRCh37 exome annotations, one for GRCh37 genome annotations, and one for each of the GRCh38 annotation chromosomes)
- Refactor Mongoose model and schema typing
- Update annotateGnomad tests

themisteryang committed:
Merge branch 'develop' into kevin/gnomAD_annotations_mongoDB

themisteryang committed:
Update node.yml and docker-compose.staging.yaml

- Update node.yml workflow file to make use of the MONGO_CONNECTION_STRING secret
- Update MONGO_CONNECTION_STRING environment variable and remove obsolete mongo container in the docker-compose.staging.yaml file
- Minor tweak to the tooltip text for the gnomAD_AF column header

themisteryang committed:
Merge branch 'kevin/gnomAD_AF' into kevin/gnomAD_annotations_mongoDB

themisteryang committed:
Merge branch 'develop' into kevin/gnomAD_annotations_mongoDB

themisteryang commented:
@jennyziyi-xu The OSMP server workflow diagram (osmp_2.png - #296) will eventually need to be updated to denote support for gnomAD GRCh38 annotations.

jennyziyi-xu commented:
> @jennyziyi-xu The OSMP server workflow diagram (osmp_2.png - #296) will eventually need to be updated to denote support for gnomAD GRCh38 annotations.

Ok that's a good point.

jennyziyi-xu commented on server/src/resolvers/getVariantsResolver/utils/annotateGnomad.ts:
```diff
@@ -1,67 +1,87 @@
         gnomadHom: nhomalt,
+        // Ideally, the cdna value should come from the CADD annotations (if available),
+        // but it can also be determined using the values from gnomAD as a fallback
         cdna:
           r.variant.info?.cdna && r.variant.info?.cdna !== 'NA'
             ? r.variant.info?.cdna
             : cdna
             ? `c.${cdna}${ref}>${alt}`
             : 'NA',
+        transcript:
+          r.variant.info?.transcript && r.variant.info?.transcript !== 'NA'
+            ? r.variant.info?.transcript
+            : transcript ?? '',
```
According to the Columns Mapping in the[ Gantt Chart ](https://docs.google.com/spreadsheets/d/1eztLBGeqjYj3IafpPxhZo5mbvz6Xzw0LbP_9_HpxeHA/edit#gid=2130371978), `cdna` and `transcript` would come from "CADD". Should we use gnomAD as fallback or remove it to avoid confusion? We haven't verified whether the values for "cdna" and "transcript" make sense. Especially for "transcript", we had extensive conversation regarding which "transcript" value to report with CADD annotations.

jennyziyi-xu commented on server/src/resolvers/getVariantsResolver/utils/annotateGnomad.ts:
```diff
@@ -1,67 +1,87 @@
 const annotateGnomad = timeitAsync('annotateGnomad')(
-  async (queryResponse: VariantQueryDataResult[]): Promise<VariantQueryDataResult[]> => {
+  async (
+    assemblyId: string,
+    position: string,
+    queryResponse: VariantQueryDataResult[]
+  ): Promise<VariantQueryDataResult[]> => {
+    const resolvedAssemblyId = resolveAssembly(assemblyId);
+    const { chromosome } = resolveChromosome(position);
     const annotationCoordinates = getCoordinates(queryResponse);
+    const GnomadAnnotationModel =
+      resolvedAssemblyId === 'GRCh37'
+        ? GnomadGRCh37AnnotationModel
+        : GnomadGRCh38AnnotationModels[chromosome];
 
-    const annotations = await GnomadAnnotationModel.getAnnotations(
-      annotationCoordinates,
-      'gnomAD_GRCh37'
-    );
+    const annotations = await GnomadAnnotationModel.getAnnotations(annotationCoordinates);
 
     return annotate(queryResponse, annotations);
```
We could add some error checking here: `if annotations.status === "fulfilled" return annotate(queryResponse....)`

(in a similar way that CADD annotations do error checking in `getVariantsResolver.ts.`

themisteryang committed:
Remove gnomAD annotations fallback for the cdna and transcript values

themisteryang committed:
Set the value of gnomAD homo count (gnomadHom) to 0 if the variant is not present in gnomAD

themisteryang committed:
Skip the CADD and gnomAD annotations steps if there are no variants to annotate

themisteryang committed:
Update error handling for variant annotation

- Refactor error handling for CADD annotation
- Add error handling for gnomAD annotation

themisteryang committed:
Merge branch 'develop' into kevin/gnomAD_annotations_mongoDB

frewmack reviewed (approved):
Looks good to me! I wasn't able to test locally with our mongodb instance, so we can iron it out in staging.

themisteryang commented:
> Looks good to me! I wasn't able to test locally with our mongodb instance, so we can iron it out in staging.

If you'd like to test the gnomAD GRCh38 annotations locally, you can try a gene from the X chromosome (e.g., PIGA)! The annotations have been uploaded and indexed for it on our local/test MongoDB instance. Chromosomes 1-8 have also been uploaded (but they haven't been indexed yet).

jennyziyi-xu reviewed (approved):
LGTM!
